src/main/rx/ghst.c

```diff
@@ -0,0 +1,259 @@
+
+STATIC_UNIT_TESTED bool ghstFrameDone = false;
+STATIC_UNIT_TESTED ghstFrame_t ghstFrame;
+STATIC_UNIT_TESTED ghstFrame_t ghstChannelDataFrame;
+STATIC_UNIT_TESTED uint32_t ghstChannelData[GHST_MAX_NUM_CHANNELS];
+
+static serialPort_t *serialPort;
+static timeUs_t ghstFrameStartAtUs = 0;
+static uint8_t telemetryBuf[GHST_FRAME_SIZE_MAX];
+static uint8_t telemetryBufLen = 0;
+
+static timeUs_t lastRcFrameTimeUs = 0;
+
+/* GHST Protocol
+ * Ghost uses 420k baud single-wire, half duplex connection, connected to a FC UART 'Tx' pin
+ * Each control packet is interleaved with one or more corresponding downlink packets
+ * 
+ * Uplink packet format (Control packets)
+ * <Addr><Len><Type><Payload><CRC>
+ *
+ * Addr:        u8      Destination address
+ * Len          u8      Length includes the packet ID, but not the CRC
+ * CRC          u8
+ * 
+ * Ghost packets are designed to be as short as possible, for minimum latency.
+ * 
+ * Note that the GHST protocol does not handle, itself, failsafe conditions. Packets are passed from 
+ * the Ghost receiver to Betaflight as and when they arrive. Betaflight itself is responsible for
+ * determining when a failsafe is necessary based on dropped packets. 
+ * 
+  */
+
+#define GHST_FRAME_LENGTH_ADDRESS       1
+#define GHST_FRAME_LENGTH_FRAMELENGTH   1
+#define GHST_FRAME_LENGTH_TYPE_CRC      1
+
+STATIC_UNIT_TESTED uint8_t ghstFrameCRC(void)
+{
+    // CRC includes type and payload
+    uint8_t crc = crc8_dvb_s2(0, ghstFrame.frame.type);
+    for (int i = 0; i < ghstFrame.frame.len - GHST_FRAME_LENGTH_TYPE_CRC - 1; ++i) {
+        crc = crc8_dvb_s2(crc, ghstFrame.frame.payload[i]);
+    }
+    return crc;
+}
+
+// Receive ISR callback, called back from serial port
+STATIC_UNIT_TESTED void ghstDataReceive(uint16_t c, void *data)
+{
+    UNUSED(data);
+
+    static uint8_t ghstFrameIdx = 0;
+    const timeUs_t currentTimeUs = microsISR();
+
+    if (cmpTimeUs(currentTimeUs, ghstFrameStartAtUs) > GHST_MAX_FRAME_TIME_US) {
+        // Character received after the max. frame time, assume that this is a new frame
+        ghstFrameIdx = 0;
+    }
+
+    if (ghstFrameIdx == 0) {
+        // timestamp the start of the frame, to allow us to detect frame sync issues
+        ghstFrameStartAtUs = currentTimeUs;
+    }
+
+    // assume frame is 5 bytes long until we have received the frame length
+    // full frame length includes the length of the address and framelength fields
+    const int fullFrameLength = ghstFrameIdx < 3 ? 5 : ghstFrame.frame.len + GHST_FRAME_LENGTH_ADDRESS + GHST_FRAME_LENGTH_FRAMELENGTH;
+
+    if (ghstFrameIdx < fullFrameLength) {
+        ghstFrame.bytes[ghstFrameIdx++] = (uint8_t)c;
+        if (ghstFrameIdx >= fullFrameLength) {                  // buffer is full?
+            ghstFrameIdx = 0;
+            const uint8_t crc = ghstFrameCRC();
+            if (crc == ghstFrame.bytes[fullFrameLength - 1]) {
+                switch (ghstFrame.frame.type)
+                {
```

Comment:
Please follow the coding style guide for formatting. (https://github.com/betaflight/betaflight/blob/master/docs/development/CodingStyle.md#curly-braces)

Reply:
No problem, change incoming